### PR TITLE
Memory64Lowering: Handle -1 return value from memory.grow

### PR DESCRIPTION
**Pull request by sbc100**
This edge case make the lowering a little more tricky.

### CODE REVIEW

File: src/passes/Memory64Lowering.cpp [hunk 1]
```diff
@@ -79,9 +79,31 @@ struct Memory64Lowering : public WalkerPass<PostWalker<Memory64Lowering>> {
...
+      // MemoryGrow returns -1 in the case of failure
+      // so we need to handle that
+      // local.set $tmp memory.grow X
+      // local.get
+      // i32.const -1
+      // if i32.eq
+      //   i64.const -1
+      // else
+      //   local.get
+      //   i32.extend
```
Comment (tlively) [inline src/passes/Memory64Lowering.cpp:91]: Can you structure this example code the same way Binaryen would print it? You can also make this slightly shorter by using a `local.tee` at the beginning if you want.

File: src/passes/Memory64Lowering.cpp [hunk 2]
```diff
@@ -79,9 +79,31 @@ struct Memory64Lowering : public WalkerPass<PostWalker<Memory64Lowering>> {
     if (memory->is64()) {
       wrapAddress64(curr->delta, curr->memory);
       auto* size = static_cast<Expression*>(curr);
-      extendAddress64(size, curr->memory);
+      // MemoryGrow returns -1 in the case of failure
+      // so we need to handle that
```
Comment (kripken) [inline src/passes/Memory64Lowering.cpp:83]:
```suggestion
      // so we need to handle that, which we do using this:
```

File: src/passes/Memory64Lowering.cpp (hunk 1 above)
Comment (kripken) [inline src/passes/Memory64Lowering.cpp:91]:
```suggestion
      // (local.set $tmp (memory.grow X))
      // (if
      //   (i32.eq
      //     (local.get $tmp)
      //     (i32.const -1)
      //   )
      //   (then
      //     (i64.const -1)
      //   )
      //   (else
      //     (i32.extend
      //       (local.get $tmp)
      //     )
      //   )
      // )
```
(might as well describe it in structured form, as that is what the code below emits)

Comment (sbc100) [reply src/passes/Memory64Lowering.cpp:91]: Is that true for the example code or also for the `builder` code?

Comment (sbc100) [reply src/passes/Memory64Lowering.cpp:91]: I actually don't know how binaryen would print this.. can you help me out?

**Review by kripken — COMMENTED** [review]
IIUC this is doing "if we get i32(-1), return i64(-1), else return u64(result)"? Can we not do "s64(result)"?

1. If the result is -1 then sign-extending it works properly.
2. If the result is the number of pages then sign-extending may be wrong if the sign bit was set. But can it be set? That would imply we allocated `2^31` pages of size `2^16` which is 47 bits... surely that is over the limit? :smile:

File: src/passes/Memory64Lowering.cpp (hunk 1 above)
Comment (sbc100) [reply src/passes/Memory64Lowering.cpp:91]: Actually I think I got it..

Comment (kripken) [reply src/passes/Memory64Lowering.cpp:91]: > Can you structure this example code the same way Binaryen would print it?

Heh I was just writing that out in a parallel comment, see below.

Comment (kripken) [reply src/passes/Memory64Lowering.cpp:91]: Using a `local.tee` makes sense but the optimizer will do it. I'm not sure if we run the optimizer before or after this pass though.

**Comment by sbc100** [comment]
> IIUC this is doing "if we get i32(-1), return i64(-1), else return u64(result)"? Can we not do "s64(result)"?

I guess that could work?

> 
> 1. If the result is -1 then sign-extending it works properly.
> 2. If the result is the number of pages then sign-extending may be wrong if the sign bit was set. But can it be set? That would imply we allocated `2^31` pages of size `2^16` which is 47 bits... surely that is over the limit? 😄

If we ever get variable sized pages I guess that would break?

**Comment by sbc100** [comment]
> > IIUC this is doing "if we get i32(-1), return i64(-1), else return u64(result)"? Can we not do "s64(result)"?
> 
> I guess that could work?
> 
> > 1. If the result is -1 then sign-extending it works properly.
> > 2. If the result is the number of pages then sign-extending may be wrong if the sign bit was set. But can it be set? That would imply we allocated `2^31` pages of size `2^16` which is 47 bits... surely that is over the limit? 😄
> 
> If we ever get variable sized pages I guess that would break?

Should we keep it simple and use signed extension in this case rather than this complex code block?  At the risk of breaking if we ever gets byte-sized pages?

**Comment by kripken** [comment]
Oh, good point, yeah, if we use variable-sized pages then we might have 1-byte pages... In which case `2^31` is not that much at all!

Let's keep the code you've already written, given that. It's not that complicated and it is future-proof.

**Comment by kripken** [comment]
Wait, how would byte-sized pages work? It seems like `memory.grow` would need to start returning an `i64`..?

**Comment by sbc100** [comment]
> Wait, how would byte-sized pages work? It seems like `memory.grow` would need to start returning an `i64`..?

Good point.  We should probably raise this on the variable page size repo.

**Comment by sbc100** [comment]
> Wait, how would byte-sized pages work? It seems like `memory.grow` would need to start returning an `i64`..?

Its only the special -1 value (0xffffffff) that is ear marked as an error... so this would just mean that under wasm32 you would only be able to grow your single-byte page-size memory to MAX_INT32-1 rather than MAX_INT32.  That doesn't seem so bad does it?

**Comment by kripken** [comment]
Noted in an upstream issue: https://github.com/WebAssembly/custom-page-sizes/issues/28

Yeah, maybe the idea is that when the page size is small you only use wasm32? But it seems like in general there are use cases that could involve wasm64.

**Review by kripken — APPROVED** [review]

**Comment by sbc100** [comment]
> Noted in an upstream issue: [WebAssembly/custom-page-sizes#28](https://github.com/WebAssembly/custom-page-sizes/issues/28)
> 
> Yeah, maybe the idea is that when the page size is small you only use wasm32? But it seems like in general there are use cases that could involve wasm64.

With a 64-bit memory, memory.grow takes i64 and returns i64

**Comment by kripken** [comment]
Oh, right, the issue is that this is for the lowering pass where the input is "unnaturally" an i32. I'll close that issue then.

**Pull request merged**